modules/network/src/main/java/org/apache/ignite/internal/network/scalecube/ScaleCubeTopologyService.java

```diff
@@ -187,8 +190,9 @@ public ClusterNode getByAddress(NetworkAddress addr) {
 
     /** {@inheritDoc} */
     @Override
-    public ClusterNode getByConsistentId(String consistentId) {
-        return consistentIdToMemberMap.get(consistentId);
+    public @Nullable ClusterNode getByConsistentId(String consistentId) {
+        Map<UUID, ClusterNode> nodes = consistentIdToMemberMap.get(consistentId);
+        return nodes != null ? nodes.values().iterator().next() : null;
```

Comment:
It seems that, even if node A' (having same name as node A) is not in the logical topology, it can still be returned here instead of node A. It seems that we look up a node by name in the following cases:

1. In `ClusterInitializer`, but this should go away when fixing the race I commented about
2. In 'business code' which only makes sense for a node in the logical topology

So my suggestion is to listen to 'node joined the logical topology' events and only add a node to this map when it gets added to the LT, not to the PT.

Reply:
I think we can achieve this at least in one of the following ways:

1. Extract LogicalTopology, listener and whatever is needed to a module like `logical-topology-api` and use it in the `network` module
2. Introduce an interface `JoinedNodes` having methods `joined()` and `left()` in `network-api`, implement it in `ScaleCubeTopologyService` and call its methods from `LogicalTopologyImpl` (here, logical topology would depend on physical topology)
3. Same as 2, but some component external to both logical topology and cluster service (maybe `IgniteImpl` itself) installs a listener on LogicalTopology, and this listener calls `JoinedNodes`' methods

Item 2 seems to be preferrable to me.